### PR TITLE
Protect QgsMapLayerRegistry for multithreaded access

### PR DESCRIPTION
**Pull request by m-kuhn**

### CODE REVIEW

**Comment by m-kuhn** [comment]
I wonder if this actually solves the problem or if it just shifts it.

In the end, what can you do safely on a rendering thread with a pointer to a layer that might get deleted anytime?

**Comment by m-kuhn** [comment]
Ping @wonder-sk @nyalldawson @jef-n

File: src/core/qgsmaplayerregistry.cpp [hunk 1]
```diff
@@ -203,14 +220,17 @@ void QgsMapLayerRegistry::onMapLayerDeleted( QObject* obj )
 
 QMap<QString, QgsMapLayer*> QgsMapLayerRegistry::mapLayers()
 {
-  return mMapLayers;
+  QReadLocker lock( &mRegistryLock );
+
+  QMap<QString, QgsMapLayer*> mapLayers = mMapLayers;
+  mapLayers.detach();
+  return mapLayers;
```
Comment (nyalldawson) [inline src/core/qgsmaplayerregistry.cpp]: Would this be needed? Isn't the first thing the container does before a modification is detach anyway?

Comment (m-kuhn) [inline src/core/qgsmaplayerregistry.cpp]: Is detach thread safe?

I.e. what happens if thread A removes a layer and detaches while thread B is iterating over the container, is there no chance for a race condition?

Comment (wonder-sk) [inline src/core/qgsmaplayerregistry.cpp]: Implicit sharing in Qt uses atomic reference counting, so the detach() call should be safe, but it is not needed in the first place because it would be called automatically when first needed...

http://doc.qt.io/qt-4.8/threads-modules.html#threads-and-implicitly-shared-classes

**Comment by wonder-sk** [comment]
Hi Matthias

Does the PR solve any particular problem or is it meant as an improvement for the future?

As you mention, working with layers in worker threads (for layers that may be used elsewhere) is dangerous - not only because they can be deleted, but also because they can change unpredictably at any time or be in a completely broken state while main thread modifies them. That's why map rendering copies everything it needs from the layers before rendering threads are started, so that race conditions are avoided.

I am not sure adding locks in map layer registry is a right path... It is difficult to get locking right and then also to keep all the implications in the head without introduction of subtle race conditions that are impossible to debug...

File: src/core/qgsmaplayerregistry.cpp (hunk 1 above)
Comment (m-kuhn) [inline src/core/qgsmaplayerregistry.cpp]: Hmm... It doesn't say thread-safe, it says

> They are fully reentrant.

And evenmore

> Note that atomic reference counting does not guarantee thread-safety. Proper locking should be used when sharing an instance of an implicitly shared class between threads. 

So "when first needed" may potentially be exactly when you are iterating. Or did I get that wrong?

File: src/core/qgsmaplayerregistry.cpp [hunk 2]
```diff
@@ -151,7 +164,9 @@ void QgsMapLayerRegistry::removeMapLayers( const QList<QgsMapLayer*>& layers )
     QString myId( lyr->id() );
     emit layerWillBeRemoved( myId );
     emit layerWillBeRemoved( lyr );
+    mRegistryLock.lockForWrite();
     mMapLayers.remove( myId );
+    mRegistryLock.unlock();
```
Comment (wonder-sk) [inline src/core/qgsmaplayerregistry.cpp]: Emitting signal layerWillBeRemoved() from a thread may cause all sorts of trouble here:
- if a slot has queued connection - the layer will be deleted before the signal is delivered
- if a slot has direct connection - suddenly a code that was not meant to be called from thread will be run in worker thread

The same applies to layersAdded() signal I believe...

**Comment by m-kuhn** [comment]
> Does the PR solve any particular problem or is it meant as an improvement for the future?

An extra layer of security which I'm not sure it's really the appropriate approach.
It's a whole can of worms there.

The other approach would be a policy: don't use the QgsMapLayerRegistry outside the main thread. But that may be hard to do e.g. for expressions referencing other layers.

Do you have an idea how that can be solved?

File: src/core/qgsmaplayerregistry.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/core/qgsmaplayerregistry.cpp]: I don't think worker threads should modify the registry.
We could insert a Q_ASSERT for this.

File: src/core/qgsmaplayerregistry.cpp (hunk 1 above)
Comment (wonder-sk) [inline src/core/qgsmaplayerregistry.cpp]: As far as I understand that, the fact that you have an implicitly shared class does not guarantee thread safety. E.g. if you have one instance and two references to it, one in each thread, implicit sharing will not help - you still need locking. But if you have two instances (sharing the same data internally), one in each thread, then you are fine, because whenever needed, detach() will be called - and the instances will stop sharing the data.

Comment (m-kuhn) [inline src/core/qgsmaplayerregistry.cpp]: Yeah, I had the same suspicion, I just haven't found any documentation saying that the call to `detach()` is threadsafe.

**Comment by wonder-sk** [comment]
> It's a whole can of worms there.

Unfortunately...

For expressions referencing other layers... that's quite a tough one - I guess it depends on what is possible to use from the other layers. If it is "just" access to features, it is possible to get hold of "feature source" for that layer (just like it is done for rendering).

**Comment by m-kuhn** [comment]
Two ways to get that:

  * Prepare all expressions when the feature iterator is created so they can get hold of the source while the worker thread has not yet started
  * Synchronize all the way from retrieving from the map layer registry to the complete feature source, so it can be done on request

**Comment by wonder-sk** [comment]
I would go for the first option - the synchronization all the way sounds like a lot of hassle and more error prone. One would need to add thread safety to map layer classes too, which would be a lot of work to cover everything (covering just minimal requirements for expressions would not be very systematic).

**Comment by m-kuhn** [comment]
So that basically involves that only static layer ids can be operated on within expressions. But it could be worse.
And improvements to the expression functions to let them execute their own preparation code (that's a nice improvement anyway)

**Comment by m-kuhn** [comment]
... and it will not be possible to use functions requiring layers inside `eval`

**Comment by m-kuhn** [comment]
Closing this, since there is much more to it than that.
Thanks for the discussion @wonder-sk
